### ISSUE TITLE
Unrestricted length of VAT Slot IDs

### ISSUE THREAD

**Issue by andrey-kuprianov** [Agoric/agoric-sdk#3242]
_Surfaced from @informalsystems audit of Agoric/agoric-sdk/SwingSet hash 9feea16732944e01749feedc7f6fb024af4fc70d_

_Notice: As agreed with Agoric, we performed the analysis under the assumption of possible bugs in the liveslots implementation. In that case, userspace vat code together with the liveslots layer operate as a whole against the kernel, which is the necessary condition for triggering this issue. Correctly functioning liveslots implementation should prevent this issue from being exploitable. We will perform assessment of liveslots correctness in the subsequent phases of our analysis._

Currently the kernel accepts VAT Slot IDs as BigInts. E.g., in the function [parseVatSlot](https://github.com/Agoric/agoric-sdk/blob/9feea16732944e01749feedc7f6fb024af4fc70d/packages/SwingSet/src/parseVatSlots.js#L36):

```js
  if (delim > 0) {
    assert(type === 'object' && allocatedByVat, X`invalid vatSlot ${s}`);
    virtual = true;
    id = Nat(BigInt(idSuffix.substr(0, delim)));
    subid = Nat(BigInt(idSuffix.slice(delim + 1)));
  } else {
    id = Nat(BigInt(idSuffix));
  }
```

 As BigInts don't have any limit, it should be possible to kill the kernel with a single syscall, by passing a very-very large number, that would eat all the memory when trying to convert from a string representation into BigInt representation.

 **Recommendation**: Restrict the size of VAT slot IDs to some sufficiently large number, e.g. to 32 bits. As an easier to enforce measure, the string representation of the number could be restricted to e.g. 10 characters.

**Comment by zarutian** [comment]
> **Recommendation**: Restrict the size of VAT slot IDs to some sufficiently large number, e.g. to 32 bits. As an easier to enforce measure, the string representation of the number could be restricted to e.g. 10 characters.

32 bit numbers for slot ids might turn out to be insufficient. 128 bits should be enough. /me states fist at IPv4 :-)

**Comment by warner** [comment]
The best place to limit this might be on the way in from the child process (the `xsnap` worker, in which the vat code lives), via an upper limit on the netstring `length` field. Anything which could hurt the kernel would probably not fit in that process, or could be prevented from arriving from that process in the first place.

**Comment by FUDCo** [comment]
Brian and I had a long discussion about this and came to the following:

- In Liveslots, we will add asserts to guard the call sites of the syscalls that send arbitrary user-originated capdata (`syscall.send`, `syscall.resolve`, and `syscall.callNow`) - this is 4 or 5 call sites total.
- The asserts will check the string length of the `body` field and the array length of the `slots` field.
- We will _not_ check the contents of the `slots` array since its elements are all Liveslots-generated vrefs, hence the vref strings themselves cannot be made unduly long (the concern that opened this issue not withstanding) without generating an enormous number of export objects which would in turn require an enormous expenditure of computation that would be limited by metering.  (The concern raised by the original issue about bigints is not germane since the bigints themselves are never accessible to user code.)
- We will presume (for now!) that all raw (i.e., non-Liveslots) vats are under our direct control and do not need this protection applied since they won't contain potentially malicious user code.
- As a pragmatic matter, we will presume (once again, for now) that Liveslots successfully confines user code such that we are not going to immediately invest in additional limiting machinery to protect the kernel from the Liveslots or worker process supervisor code being compromised via an exploit.
- Not withstanding that, longer term we do intend to add a length check in the netstring decoder that reads the worker process to kernel process message pipe so that even a compromised worker process cannot overflow the kernel's memory space by sending excessively long syscall requests.
- As a first cut, we picked limits of 10 million characters for body strings and 10,000 elements for slot arrays.  These limits will be initialized from constants in `liveslots.js` but alterable via a test hook to make testing the limit code practical.
- Once the refactoring of the Liveslots vat options #5619 is done, we can add the syscall capdata size limit values to the set of things that can be set in the future via governance, but the need for this is not so pressing as to warrant adding further ad hoc parameter setting code right now.

**Comment by mhofman** [comment]
I just found out that the netstring limit does not apply to xnsap workers. New issue: https://github.com/Agoric/agoric-sdk/issues/6150
